cache/store.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2025 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"bytes"
+	"sort"
+	"sync"
+
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+type store struct {
+	mu        sync.RWMutex
+	kvs       map[string]*mvccpb.KeyValue
+	latestRev int64
+}
+
+func newStore() *store {
+	return &store{kvs: make(map[string]*mvccpb.KeyValue)}
+}
+
+func (s *store) initializeStoreState(kvs []*mvccpb.KeyValue, rev int64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for _, kv := range kvs {
+		s.kvs[string(kv.Key)] = kv
+	}
+	s.latestRev = rev
+}
+
+func (s *store) apply(events []*clientv3.Event) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, ev := range events {
+		if ev.Type == clientv3.EventTypeDelete {
+			delete(s.kvs, string(ev.Kv.Key))
+		} else {
+			s.kvs[string(ev.Kv.Key)] = ev.Kv
+		}
+		if ev.Kv.ModRevision > s.latestRev {
+			s.latestRev = ev.Kv.ModRevision
+		}
+	}
+}
+
+func (s *store) selectKVs(pred KeyPredicate) ([]*mvccpb.KeyValue, int64) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	var out []*mvccpb.KeyValue
+	for _, kv := range s.kvs {
```

Comment:
What happens when someone just want to get single key, you iterate all keys to just find one instead of map lookup?